src/Raven.Client/Documents/Identity/AsyncMultiTypeHiLoIdGenerator.cs

```diff
@@ -30,27 +29,35 @@ public AsyncMultiTypeHiLoIdGenerator(DocumentStore store, string dbName, Documen
             Conventions = conventions;
         }
 
-        public Task<string> GenerateDocumentIdAsync(object entity)
+        public async Task<string> GenerateDocumentIdAsync(object entity)
         {
             var typeTagName = Conventions.GetCollectionName(entity);
             if (string.IsNullOrEmpty(typeTagName)) //ignore empty tags
             {
-                return NullStringCompletedTask;
+                return null;
             }
             var tag = Conventions.TransformTypeCollectionNameToDocumentIdPrefix(typeTagName);
             if (_idGeneratorsByTag.TryGetValue(tag, out var value))
-                return value.GenerateDocumentIdAsync(entity);
+            {
+                return await value.GenerateDocumentIdAsync(entity).ConfigureAwait(false);
+            }
 
-            lock (_generatorLock)
+            await _generatorLock.WaitAsync().ConfigureAwait(false);
+
+            try
             {
                 if (_idGeneratorsByTag.TryGetValue(tag, out value))
-                    return value.GenerateDocumentIdAsync(entity);
+                    return await value.GenerateDocumentIdAsync(entity).ConfigureAwait(false);
```

Comment:
This is likely a problem. 
You are doing the expensive ( network call wait) _while_ you are holding the lock.
This needs to be moved outside the lock

Reply:
done

Reply:
_idGeneratorsByTag is a ConcurrentDictionary can't we use GetOrAdd?

Reply:
no, the `lock` there is fine